File: Codechef/May Challenge 2021/THOUSES.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'
#define int unsigned long long 
#define __ ios_base::sync_with_stdio(false);cin.tie(NULL);

const int maxn = 300007;
vector<int> graph[maxn];
const int mod = 1e9+7;
vector<int> val(maxn);

void dfs(int u,int p = -1){
    int c = 0;
    for(auto v:graph[u]){
        if(v==p)continue;
        dfs(v,u);
        c++;
    }
    if(!c)return;
    vector<int> childrens;
    for(auto v:graph[u]){
        if(v==p)continue;
        childrens.push_back(val[v]);
    }

    sort(childrens.rbegin(),childrens.rend());
    for(int i = 1;i<=c;i++){
        if(__builtin_uaddll_overflow(val[u],childrens[i-1]*i,&val[u])){
            cout<<"OVERFLOW"<<endl;
            return;
        }
    }
}
signed main(){__
	int t= 1,n,u,v,x;
    cin>>t;
    while(t--){
        cin>>n>>x;
        for(int i = 0;i<n;i++)graph[i].clear(),val[i] = 1;
        for(int i = 0;i<n-1;i++){
            cin>>u>>v;
            u--,v--;
            graph[u].push_back(v);
            graph[v].push_back(u);
        }
        dfs(0);
        cout<<((val[0]%mod)*(x%mod))%mod<<endl;
    }
    return 0;
}  

File: Codechef/May Challenge 2021/THOUSES.py
import math
val = []
graph = []
def dfs(u,p):
    c = 0    
    for v in graph[u]:
        if v==p:
            continue
        dfs(v,u)
        c = c+1
    childrens =[]
    for v in graph[u]:
        if v==p:
            continue
        childrens.append(val[v])
    childrens.sort(reverse=True)
    for i in range(1,c+1):
        x = val[u]
        val[u] = x+(childrens[i-1]*i)

mod = 1e9+7
t=int(input())
while t:
    n,x=input().split()
    n=int(n)
    x=int(x)
    val = []
    graph = []
    for i in range(0,n):
        graph.append([]),val.append(x) 
    
    for i in range(0,n-1):
        a,b=input().split()
        graph[int(a)-1].append(int(b)-1)
        graph[int(b)-1].append(int(a)-1)
    dfs(0,-1)
    print(math.floor(val[0]%mod))
    t-=1